### PR TITLE
INS-91: Add record lifeline index to storage

### PR DESCRIPTION
**Pull request by andreyromancev**
Closes INS-91

**- What I did**
Added lifeline index

**- How to verify it**
Run tests

### CODE REVIEW

File: ledger/storage/leveldb/leveldb.go [hunk 1]
```diff
@@ -33,9 +33,14 @@ const (
 // LevelLedger represents ledger's LevelDB storage.
 type LevelLedger struct {
 	// LDB contains LevelDB database instance.
-	LDB *leveldb.DB
+	ldb *leveldb.DB
 }
 
+const (
+	scopeIDRecord byte = iota
```
Comment (golangci) [inline ledger/storage/leveldb/leveldb.go]: `scopeIDRecord` is unused

File: ledger/storage/leveldb/leveldb_test.go [hunk 2]
```diff
@@ -3,8 +3,44 @@ package leveldb
...
+	absPath, err := filepath.Abs(dbDirPath)
+	if err == nil {
+		os.RemoveAll(absPath)
+	}
 	os.Exit(m.Run())
 }
+
+func TestGetIndexOnEmptyDataReturnsNotFound(t *testing.T) {
+	ledger, err := InitDB()
+	defer ledger.Close()
```
Comment (golangci) [inline ledger/storage/leveldb/leveldb_test.go]: should check returned error before deferring ledger.Close()

File: ledger/storage/leveldb/leveldb_test.go [hunk 3]
```diff
@@ -3,8 +3,44 @@ package leveldb
...
+	assert.Equal(t, err, nil)
+
+	index, isFound := ledger.GetIndex(record.ID{1})
+	assert.Equal(t, isFound, false)
+	assert.Equal(t, index, record.LifelineIndex{})
+}
+
+func TestSetIndexStoresDataInDB(t *testing.T) {
+	ledger, err := InitDB()
+	defer ledger.Close()
```
Comment (golangci) [inline ledger/storage/leveldb/leveldb_test.go]: should check returned error before deferring ledger.Close()

File: ledger/record/index_serialize.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"bytes"
+
+	"golang.org/x/crypto/sha3"
+	"github.com/ugorji/go/codec"
+)
+
+// HashLifelineIndex returns 28 bytes of SHA3 hash
+func HashLifelineIndex(buf []byte) Hash {
+	return sha3.Sum224(buf)
```
Comment (nordicdyno) [inline ledger/record/index_serialize.go]: it's a better use hash.SHA3hash224 implementation, according to the single responsibility principle.

i.e. to implement WriteHash method on LifelineIndex

File: ledger/record/record_serialize.go [hunk 5]
```diff
@@ -20,26 +20,25 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
-
-	"github.com/ugorji/go/codec"
+		"github.com/ugorji/go/codec"
```
Comment (nordicdyno) [inline ledger/record/record_serialize.go]: indent issue (you should use go fmt)

File: ledger/record/record_serialize.go [hunk 6]
```diff
@@ -20,26 +20,25 @@ import (
...
 	"golang.org/x/crypto/sha3"
 )
 
-// Raw struct contains raw serialized record.
+// RawRecord struct contains raw serialized record.
```
Comment (nordicdyno) [inline ledger/record/record_serialize.go]: why? we already have record package, so record.Raw is pretty specific name

File: ledger/storage/leveldb/leveldb.go [hunk 7]
```diff
@@ -92,16 +96,36 @@ func InitDB() (*LevelLedger, error) {
...
 
-// Set stores record in leveldb
-func (ll *LevelLedger) Set(record record.Record) error {
+// SetRecord stores record in leveldb
+func (ll *LevelLedger) SetRecord(record record.Record) error {
 	return nil
 }
+
+// GetIndex fetches lifeline index from leveldb (records and lifeline indexes have the same id, but different scopes)
+func (ll *LevelLedger) GetIndex(id record.ID) (index record.LifelineIndex, found bool) {
```
Comment (nordicdyno) [inline ledger/storage/leveldb/leveldb.go]: don't use named return values (some rare cases are ok, but this language feature is mostly for very short helper methods, and in production code, we should avoid it as much as possible)

**Pull request closed without merging**